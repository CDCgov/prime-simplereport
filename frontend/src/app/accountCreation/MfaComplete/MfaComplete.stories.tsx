import { StoryFn, Meta } from "@storybook/react";

import { MfaComplete } from "./MfaComplete";

export default {
  title: "App/Account set up/Step 3c: MFA complete",
  component: MfaComplete,
  argTypes: {},
} as Meta;

const Template: StoryFn = (args) => <MfaComplete {...args} />;

export const Default = Template.bind({});
Default.args = {};
