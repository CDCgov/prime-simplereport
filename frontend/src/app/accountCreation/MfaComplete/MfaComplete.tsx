import { Card } from "../../commonComponents/Card/Card";
import { CardBackground } from "../../commonComponents/CardBackground/CardBackground";
import Button from "../../commonComponents/Button/Button";
import iconSprite from "../../../../node_modules/uswds/dist/img/sprite.svg";

export const MfaComplete = () => {
  return (
    <CardBackground>
      <Card logo>
        <div className="display-flex flex-justify-center margin-top-2">
          <svg
            className="usa-icon usa-icon--size-8 text-success"
            aria-hidden="true"
            focusable="false"
            role="img"
          >
            <use xlinkHref={iconSprite + "#check_circle"}></use>
          </svg>
        </div>
        <h1 className="font-ui-lg margin-top-3">Account set up complete</h1>
        <p className="margin-bottom-0">
          To start using SimpleReport, log in to your account.
        </p>
        <div className="display-flex flex-justify-center">
          <Button
            className="margin-top-3 flex-justify-center"
            label={"Continue to SimpleReport"}
            type={"submit"}
            onClick={() => (window.location.pathname = "/")}
          />
        </div>
      </Card>
    </CardBackground>
  );
};
