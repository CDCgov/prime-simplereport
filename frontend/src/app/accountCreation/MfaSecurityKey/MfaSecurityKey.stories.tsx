import { StoryFn, Meta } from "@storybook/react-webpack5";

import { getMocks } from "../../../stories/storyMocks";

import { MfaSecurityKey } from "./MfaSecurityKey";

export default {
  title: "App/Account set up/Step 3a: Security key",
  component: MfaSecurityKey,
  parameters: {
    msw: getMocks("enrollSecurityKeyMfa"),
  },
  argTypes: {},
} as Meta;

const Template: StoryFn = (args) => <MfaSecurityKey {...args} />;

export const Default = Template.bind({});
Default.args = {};
