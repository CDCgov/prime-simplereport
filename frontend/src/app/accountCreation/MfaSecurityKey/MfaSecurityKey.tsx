import React, { useEffect, useState } from "react";

import { Card } from "../../commonComponents/Card/Card";
import { CardBackground } from "../../commonComponents/CardBackground/CardBackground";
import StepIndicator from "../../commonComponents/StepIndicator";
import { accountCreationSteps } from "../../../config/constants";
import { useDocumentTitle } from "../../utils/hooks";

export const MfaSecurityKey = () => {
  useDocumentTitle("Set up authentication via security key");

  const [unsupported, setUnsupported] = useState(false);

  useEffect(() => {
    // navigator.credentials is a global object on WebAuthn-supported clients, used to access WebAuthn API
    // if the user's browser doesn't support WebAuthn, display a message telling them to use a different browser
    if (!navigator?.credentials) {
      setUnsupported(true);
    }
  }, []);

  if (unsupported) {
    return (
      <CardBackground>
        <Card logo bodyKicker="Unsupported browser">
          <p className="margin-bottom-0">
            To register a security key, please use the latest version of a
            supported browser:
          </p>
          <ol className="usa-list usa-hint font-ui-2xs">
            <li>Google Chrome</li>
            <li>Mozilla Firefox</li>
            <li>Microsoft Edge</li>
          </ol>
        </Card>
      </CardBackground>
    );
  }

  return (
    <CardBackground>
      <Card logo bodyKicker="Set up your account">
        <StepIndicator
          steps={accountCreationSteps}
          currentStepValue={"2"}
          noLabels={true}
        />
        <p className="margin-bottom-0">How to register your security key.</p>
        <ol className="usa-list usa-hint font-ui-2xs">
          <li>Continue to SimpleReport and log in.</li>
          <li>
            Find “Security Key or Biometric Authenticator” in the Okta
            multifactor authentication menu and click{" "}
            <span className="text-bold">Setup</span>.
          </li>
          <li>Follow prompts to enroll.</li>
        </ol>
        <a
          href={"/app/"}
          className="margin-x-auto display-block width-fit-content usa-button usa-button--primary"
        >
          Continue to SimpleReport
        </a>
      </Card>
    </CardBackground>
  );
};
