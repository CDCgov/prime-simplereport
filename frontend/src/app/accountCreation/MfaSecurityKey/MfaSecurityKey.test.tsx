import { render, screen, act } from "@testing-library/react";
import { MemoryRouter, Route } from "react-router";

import { MfaComplete } from "../MfaComplete/MfaComplete";

import { MfaSecurityKey } from "./MfaSecurityKey";

jest.mock("../AccountCreationApiService", () => ({
  AccountCreationApi: {
    enrollSecurityKeyMfa: () => {
      return new Promise((res) => {
        res({ activation: { challenge: "challenge", user: { id: "userId" } } });
      });
    },
    activateSecurityKeyMfa: (attestation: string, clientData: string) => {
      return new Promise((res, rej) => {
        if (attestation.length !== 0 && clientData.length !== 0) {
          res("success");
        } else {
          rej("utter failure");
        }
      });
    },
  },
}));

jest.mock("../../utils/text", () => ({
  strToBin: (str: string) => {
    return Uint8Array.of(8).fill(2);
  },
  binToStr: (bin: ArrayBuffer) => {
    return "str";
  },
}));

function create(obj: any) {
  return new Promise((cred) => {
    cred({
      response: {
        attestationObject: "attestation",
        clientDataJSON: "clientDataJSON",
      },
    });
  });
}

describe("MFA Security Key Successful", () => {
  beforeEach(() => {
    let credContainer = { create: create };
    Object.defineProperty(window.navigator, "credentials", {
      value: credContainer,
      configurable: true,
    });
  });

  it("displays the success page if security key is activated", async () => {
    // in the real flow, a user has to touch their security key to trigger the enrollment.
    // it's not possible to mock that, so instead we mock out functions as if the user
    // had touched their key and assert that they get to the success page.
    act(() => {
      render(
        <MemoryRouter initialEntries={["/enroll-security-key-mfa"]}>
          <Route path="/enroll-security-key-mfa" component={MfaSecurityKey} />
          <Route path="/success" component={MfaComplete} />
        </MemoryRouter>
      );
    });

    expect(
      await screen.findByText("Account set up complete")
    ).toBeInTheDocument();
  });
});
