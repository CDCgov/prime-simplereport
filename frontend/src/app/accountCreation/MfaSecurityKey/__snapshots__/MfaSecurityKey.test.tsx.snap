// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MFA Security Key Successful displays MFA security key step 1`] = `
<body>
  <div>
    <div
      class="card__background bg-base-lightest"
    >
      <div
        class="grid-container maxw-mobile-lg usa-section"
      >
        <div
          class="card"
        >
          <header
            class="display-flex flex-column"
          >
            <img
              alt="SimpleReport logo"
              class="flex-align-self-center maxw-card-lg width-full"
              src="simplereport-logo-color.svg"
            />
            <div
              class="border-bottom border-base-lighter margin-x-neg-3 margin-top-3"
            />
          </header>
          <p
            class="font-ui-sm text-bold margin-top-3"
          >
            Set up your account
          </p>
          <div
            class="usa-step-indicator usa-step-indicator--counters-sm usa-step-indicator--no-labels margin-y-205"
          >
            <div
              class="usa-step-indicator__segments"
              role="presentation"
            >
              <div
                aria-current="false"
                class="usa-step-indicator__segment usa-step-indicator__segment--complete"
              >
                <span
                  class="usa-step-indicator__segment-label"
                >
                  Create your password
                   
                  <span
                    class="usa-sr-only"
                  >
                    completed
                  </span>
                </span>
              </div>
              <div
                aria-current="false"
                class="usa-step-indicator__segment usa-step-indicator__segment--complete"
              >
                <span
                  class="usa-step-indicator__segment-label"
                >
                  Select your security question
                   
                  <span
                    class="usa-sr-only"
                  >
                    completed
                  </span>
                </span>
              </div>
              <div
                aria-current="true"
                class="usa-step-indicator__segment usa-step-indicator__segment--current"
              >
                <span
                  class="usa-step-indicator__segment-label"
                >
                  Set up authentication
                   
                  <span
                    class="usa-sr-only"
                  >
                    completed
                  </span>
                </span>
              </div>
            </div>
            <div
              class="usa-step-indicator__header"
            >
              <h1
                class="usa-step-indicator__heading"
              >
                <span
                  class="usa-step-indicator__heading-counter"
                >
                  <span
                    class="usa-sr-only"
                  >
                    Step
                  </span>
                  <span
                    class="usa-step-indicator__current-step margin-right-05"
                  >
                    3
                  </span>
                  <span
                    class="usa-step-indicator__total-steps"
                  >
                    of 
                    3
                  </span>
                </span>
                <span
                  class="usa-step-indicator__heading-text"
                >
                  Set up authentication
                </span>
              </h1>
            </div>
          </div>
          <p
            class="margin-bottom-0"
          >
            How to register your security key.
          </p>
          <ol
            class="usa-list usa-hint font-ui-2xs"
          >
            <li>
              Continue to SimpleReport and log in.
            </li>
            <li>
              Find “Security Key or Biometric Authenticator” in the Okta multifactor authentication menu and click
               
              <span
                class="text-bold"
              >
                Setup
              </span>
              .
            </li>
            <li>
              Follow prompts to enroll.
            </li>
          </ol>
          <a
            class="margin-x-auto display-block width-fit-content usa-button usa-button--primary"
            href="/app/"
          >
            Continue to SimpleReport
          </a>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`MFA Security Key Successful displays MFA security key step for unsupported browser 1`] = `
<body>
  <div>
    <div
      class="card__background bg-base-lightest"
    >
      <div
        class="grid-container maxw-mobile-lg usa-section"
      >
        <div
          class="card"
        >
          <header
            class="display-flex flex-column"
          >
            <img
              alt="SimpleReport logo"
              class="flex-align-self-center maxw-card-lg width-full"
              src="simplereport-logo-color.svg"
            />
            <div
              class="border-bottom border-base-lighter margin-x-neg-3 margin-top-3"
            />
          </header>
          <p
            class="font-ui-sm text-bold margin-top-3"
          >
            Unsupported browser
          </p>
          <p
            class="margin-bottom-0"
          >
            To register a security key, please use the latest version of a supported browser:
          </p>
          <ol
            class="usa-list usa-hint font-ui-2xs"
          >
            <li>
              Google Chrome
            </li>
            <li>
              Mozilla Firefox
            </li>
            <li>
              Microsoft Edge
            </li>
          </ol>
        </div>
      </div>
    </div>
  </div>
</body>
`;
